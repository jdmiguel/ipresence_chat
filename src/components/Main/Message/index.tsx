import React from 'react';
import { InView } from 'react-intersection-observer';

import { observerOptions } from '../../../helpers/utils';
import { IDisplayedMessage } from '../../../helpers/types';

interface IMessageProps {
  data: IDisplayedMessage;
  onUnreadMessage: (id: number) => void;
}

const Message: React.FC<IMessageProps> = ({
  data: {
    id,
    text,
    messageClasses,
    isUnread,
    hasIcon,
    iconClasses,
    iconName,
    date,
  },
  onUnreadMessage,
}) => (
  <InView
    {...observerOptions}
    as="div"
    className={messageClasses}
    onChange={(inView) => {
      if (inView && isUnread) {
        onUnreadMessage(id);
      }
    }}
  >
    <div className="message-content">
      <p>{text}</p>
      {hasIcon && (
        <i className={iconClasses && iconClasses}>{iconName && iconName}</i>
      )}
    </div>
    <div className="message-time">
      <p>{date}</p>
    </div>
  </InView>
);

export default Message;
