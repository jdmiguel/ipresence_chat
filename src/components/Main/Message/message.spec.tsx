import React from 'react';
import { render } from '@testing-library/react';
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils'

import Message from '.';

afterEach(() => {
  mockAllIsIntersecting(true);
})

const MESSAGE_PROPS = {
  data: {
    id: 0,
    text: 'Hi',
    messageClasses: '',
    isUnread: false,
    hasIcon: false,
    iconClasses: '',
    iconName: '',
    date: '10:00 - April 12, 2020',
  },
  onUnreadMessage: () => {}
};

describe('Component: Message', () => {
  it('should render as outcoming message and display text properly', () => {
    const { container } = render(
      <Message {...MESSAGE_PROPS} />
    );
    const messageWrapper = container.firstChild as HTMLDivElement;
    const text = messageWrapper.querySelector('.message-content > p')?.innerHTML;

    expect(messageWrapper).toMatchSnapshot();
    expect(text).toBe('Hi');
  });

  it('should render as outcoming message and as sent', () => {
    const MESSAGE_PROPS_UPDATED = {...MESSAGE_PROPS, data: {...MESSAGE_PROPS.data, hasIcon:true }}
    const { container } = render(
      <Message {...MESSAGE_PROPS_UPDATED} />
    );
    const messageWrapper = container.firstChild as HTMLDivElement;

    expect(messageWrapper).toMatchSnapshot();
  });

  it('should render as outcoming message and as received', () => {
    const MESSAGE_PROPS_UPDATED = {...MESSAGE_PROPS, data: {...MESSAGE_PROPS.data, hasIcon:true, iconName: 'done_all'}}
    const { container } = render(
      <Message {...MESSAGE_PROPS_UPDATED} />
    );
    const messageWrapper = container.firstChild as HTMLDivElement;

    expect(messageWrapper).toMatchSnapshot();
});

  it('should render as outcoming message and as read', () => {
      const MESSAGE_PROPS_UPDATED = {...MESSAGE_PROPS, data: {...MESSAGE_PROPS.data, hasIcon:true, iconClasses: 'read', iconName: 'done_all'  }}
      const { container } = render(
        <Message {...MESSAGE_PROPS_UPDATED} />
      );
      const messageWrapper = container.firstChild as HTMLDivElement;
  
      expect(messageWrapper).toMatchSnapshot();
  });

  it('should render as incoming message', () => {
    const MESSAGE_PROPS_UPDATED = {...MESSAGE_PROPS, data: {...MESSAGE_PROPS.data, messageClasses:'message incoming'}}
    const { container } = render(
      <Message {...MESSAGE_PROPS_UPDATED} />
    );
    const messageWrapper = container.firstChild as HTMLDivElement;

    expect(messageWrapper).toMatchSnapshot();
  });
});