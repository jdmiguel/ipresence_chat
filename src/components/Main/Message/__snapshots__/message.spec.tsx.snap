// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Message should render as incoming message 1`] = `
<div
  class="message incoming"
>
  <div
    class="message-content"
  >
    <p>
      Hi
    </p>
  </div>
  <div
    class="message-time"
  >
    <p>
      10:00 - April 12, 2020
    </p>
  </div>
</div>
`;

exports[`Component: Message should render as outcoming message and as read 1`] = `
<div
  class=""
>
  <div
    class="message-content"
  >
    <p>
      Hi
    </p>
    <i
      class="read"
    >
      done_all
    </i>
  </div>
  <div
    class="message-time"
  >
    <p>
      10:00 - April 12, 2020
    </p>
  </div>
</div>
`;

exports[`Component: Message should render as outcoming message and as received 1`] = `
<div
  class=""
>
  <div
    class="message-content"
  >
    <p>
      Hi
    </p>
    <i
      class=""
    >
      done_all
    </i>
  </div>
  <div
    class="message-time"
  >
    <p>
      10:00 - April 12, 2020
    </p>
  </div>
</div>
`;

exports[`Component: Message should render as outcoming message and as sent 1`] = `
<div
  class=""
>
  <div
    class="message-content"
  >
    <p>
      Hi
    </p>
    <i
      class=""
    />
  </div>
  <div
    class="message-time"
  >
    <p>
      10:00 - April 12, 2020
    </p>
  </div>
</div>
`;

exports[`Component: Message should render as outcoming message and display text properly 1`] = `
<div
  class=""
>
  <div
    class="message-content"
  >
    <p>
      Hi
    </p>
  </div>
  <div
    class="message-time"
  >
    <p>
      10:00 - April 12, 2020
    </p>
  </div>
</div>
`;
